pkg/apis/nfd/validate/validate.go

```diff
@@ -0,0 +1,206 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package validate
+
+import (
+	"fmt"
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+	k8sQuantity "k8s.io/apimachinery/pkg/api/resource"
+	k8svalidation "k8s.io/apimachinery/pkg/util/validation"
+	"k8s.io/klog/v2"
+
+	nfdv1alpha1 "sigs.k8s.io/node-feature-discovery/pkg/apis/nfd/v1alpha1"
+)
+
+var (
+	// Default error message for invalid label/annotation keys
+	ErrNSNotAllowed = fmt.Errorf("namespace is not allowed")
+	// Default error message for invalid label/annotation keys
+	ErrUnprefixedKeysNotAllowed = fmt.Errorf("unprefixed keys are disallowed")
+	// Default error for invalid taint effect
+	ErrInvalidTaintEffect = fmt.Errorf("invalid taint effect")
+	// Default error for empty taint effect
+	ErrEmptyTaintEffect = fmt.Errorf("empty taint effect")
+)
+
+func getDynamicValue(value string, features *nfdv1alpha1.Features) (string, error) {
+	// value is a string in the form of attribute.featureset.elements
+	split := strings.SplitN(value[1:], ".", 3)
+	if len(split) != 3 {
+		return "", fmt.Errorf("value %s is not in the form of '@domain.feature.element'", value)
+	}
+	featureName := split[0] + "." + split[1]
+	elementName := split[2]
+	attrFeatureSet, ok := features.Attributes[featureName]
+	if !ok {
+		return "", fmt.Errorf("feature %s not found", featureName)
+	}
+	element, ok := attrFeatureSet.Elements[elementName]
+	if !ok {
+		return "", fmt.Errorf("element %s not found on feature %s", elementName, featureName)
+	}
+	return element, nil
+}
+
+// Label validates a label key and value and returns an error if the key or
+// value is invalid.
+func Label(key, value string, features *nfdv1alpha1.Features) (string, error) {
+	//Validate label key and value
+	if err := k8svalidation.IsQualifiedName(key); len(err) > 0 {
+		return "", fmt.Errorf("invalid label key %q: %s", key, strings.Join(err, "; "))
+	}
+	// Check label namespace, filter out if ns is not whitelisted
+	ns, _ := splitNs(key)
+	if ns != nfdv1alpha1.FeatureLabelNs && ns != nfdv1alpha1.ProfileLabelNs &&
+		!strings.HasSuffix(ns, nfdv1alpha1.FeatureLabelSubNsSuffix) && !strings.HasSuffix(ns, nfdv1alpha1.ProfileLabelSubNsSuffix) {
+		// And is not empty
+		if ns == "" {
+			klog.ErrorS(fmt.Errorf("labels without namespace (prefix/) not allowed"), fmt.Sprintf("Ignoring annotation %s", key))
+			return "", ErrUnprefixedKeysNotAllowed
+		}
+		// And is not a denied namespace
+		if ns == "kubernetes.io" || strings.HasSuffix(ns, ".kubernetes.io") {
+			return "", ErrNSNotAllowed
+		}
+	}
+
+	// Validate label value
+	if strings.HasPrefix(value, "@") {
+		dynamicValue, err := getDynamicValue(value, features)
```

Comment:
I'd suggest to keep this dynamic expansion of values in nfd-master. I.e. first expand the value and the call `validate.Labe()` to check if everything looks good. This would also make the `validate` package/API more uniform, i.e. all validate functions return just an error, not a value plus an error (and also as input you don't need to give Features).